adapters/contxtful/contxtful.go

```diff
@@ -0,0 +1,405 @@
+	Params ContxtfulBidRequestParams `json:"params"`
+	BidID  string                    `json:"bidId"`
+}
+
+// ContxtfulBidRequestParams represents the parameters for a bid request
+type ContxtfulBidRequestParams struct {
+	PlacementID string `json:"placementId"`
+}
+
+// ContxtfulBidderRequest represents the bidder request information
+type ContxtfulBidderRequest struct {
+	BidderCode string `json:"bidderCode"`
+}
+
+// ContxtfulConfig represents the configuration section
+type ContxtfulConfig struct {
+	Contxtful ContxtfulConfigDetails `json:"contxtful"`
+}
+
+// ContxtfulConfigDetails represents the detailed configuration
+type ContxtfulConfigDetails struct {
+	Version  string `json:"version"`
+	Customer string `json:"customer"`
+}
+
+// BidExtensions represents the complete bid extensions structure
+type BidExtensions struct {
+	OrigBidCPM float64                  `json:"origbidcpm"`
+	OrigBidCur string                   `json:"origbidcur"`
+	Prebid     openrtb_ext.ExtBidPrebid `json:"prebid"`
+}
+
+type ContxtfulExchangeBid struct {
+	RequestID   string              `json:"requestId"`
+	CPM         float64             `json:"cpm"`
+	Currency    string              `json:"currency"`
+	Width       int                 `json:"width"`
+	Height      int                 `json:"height"`
+	CreativeID  string              `json:"creativeId"`
+	AdM         string              `json:"adm"`
+	TTL         int                 `json:"ttl"`
+	NetRevenue  bool                `json:"netRevenue"`
+	MediaType   string              `json:"mediaType"`
+	BidderCode  string              `json:"bidderCode"`
+	PlacementID string              `json:"placementId"`
+	TraceId     string              `json:"traceId,omitempty"`
+	Random      float64             `json:"random,omitempty"`
+	NURL        string              `json:"nurl"`
+	BURL        string              `json:"burl"`
+	LURL        string              `json:"lurl,omitempty"`
+	Ext         jsonutil.RawMessage `json:"ext,omitempty"`
+}
+
+type BidProcessingContext struct {
+	request        *openrtb2.BidRequest
+	requestData    *adapters.RequestData
+	customerId     string
+	bidderResponse *adapters.BidderResponse
+	errors         []error
+	// Event URL generation fields
+	version     string
+	domain      string
+	adRequestID string
+}
+
+type adapter struct {
+	endpointTemplate *template.Template
+}
+
+// Essential constants
+const (
+	BidderName     = "contxtful"
+	DefaultVersion = "v1"
+	FieldBidder    = "bidder"
+	PbsPath        = "/pbs/"
+)
+
+// Helper to safely extract bidder parameters from impression extension
+func extractBidderParams(impExt jsonutil.RawMessage) (*openrtb_ext.ExtImpContxtful, error) {
+	var bidderExt adapters.ExtImpBidder
+	if err := jsonutil.Unmarshal(impExt, &bidderExt); err != nil {
+		return nil, err
+	}
+
+	var contxtfulParams openrtb_ext.ExtImpContxtful
+	if err := jsonutil.Unmarshal(bidderExt.Bidder, &contxtfulParams); err != nil {
+		return nil, err
```

Comment:
Add supplemental JSON test for this error path. This can be done by making the `imp.ext.bidder` a type other than an object.